### PR TITLE
SockJS client for XHR-Polling 

### PR DESCRIPTION
**Pull request by fatih**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 language: go
-go: 1.3
+go: 1.4
```
Comment (cihangir) [inline .travis.yml]: better to have 1.3 too

File: kite_test.go [hunk 2]
```diff
@@ -263,3 +271,9 @@ func SquareCB(r *Request) (interface{}, error) {
 
 	return nil, nil
 }
+
+func NewKite(name, version string) *Kite {
```
Comment (cihangir) [inline kite_test.go]: NewXHRPollingKite may be? yes it is a long one but more expressive :) On the other hand if it is just a test helper you can un-export it

File: kontrol/kontrol.go [hunk 3]
```diff
@@ -20,11 +20,11 @@ import (
 const (
 	KontrolVersion = "0.0.4"
 	KitesPrefix    = "/kites"
-	TokenLeeway    = 1 * time.Minute
 )
 
 var (
 	TokenTTL    = 48 * time.Hour
+	TokenLeeway = 1 * time.Minute
```
Comment (cihangir) [inline kontrol/kontrol.go]: TIL `Leeway: (noun) room to move, grow`

File: sockjsclient/xhr.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		sessionURL: sessionURL,
+		opened:     true,
+	}, nil
+}
+
+func (x *XHRSession) ID() string {
+	return x.sessionID
+}
+
+func (x *XHRSession) Recv() (string, error) {
```
Comment (cihangir) [inline sockjsclient/xhr.go]: is `Recv()` syncronized for all consumers? Or is there a guarantee that there will be only one consumer(the place where we are calling Recv)?

File: sockjsclient/xhr.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		return msg, nil
+	}
+
+	// start to poll from the server until we receive something
+	for {
+		resp, err := x.client.Post(x.sessionURL+"/xhr", "text/plain", nil)
+		if err != nil {
+			return "", err
+		}
+		defer resp.Body.Close()
```
Comment (cihangir) [inline sockjsclient/xhr.go]: make sure response.StatusCode is http.StatusOK (according to spec section you wrote on line 31)

File: sockjsclient/xhr.go (hunk 4 above)
Comment (fatih) [inline sockjsclient/xhr.go]: Yes it's synchronized here: https://github.com/koding/kite/blob/master/client.go#L270. Basically `readloop()` is called once and every received message is processed in a separate goroutine. We have also an option to disable concurrent processing of messaging (via the config.DisableConcurrency option)

File: sockjsclient/xhr.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+			x.opened = true
+			x.mu.Unlock()
+			continue
+		case 'a':
+			// received an array of messages
+			data, err := ioutil.ReadAll(buf)
+			if err != nil {
+				return "", err
+			}
+
```
Comment (cihangir) [inline sockjsclient/xhr.go]: you can Decode all the messages with 
```
err = json.NewDecoder(buf).Decode(&message)
```
and omit the ioutill.ReadAll & json.Unmarshal

File: sockjsclient/xhr.go (hunk 5 above)
Comment (fatih) [inline sockjsclient/xhr.go]: Done.

File: sockjsclient/xhr.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	resp, err := x.client.Post(x.sessionURL+"/xhr_send", "text/plain", bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == 404 {
+		return errors.New("XHR session doesn't exists")
+	}
+
+	if resp.StatusCode != 204 {
```
Comment (cihangir) [inline sockjsclient/xhr.go]: :+1:

File: .travis.yml (hunk 1 above)
Comment (fatih) [inline .travis.yml]: Done.

**Comment by cihangir** [comment]
LGTM. Thanks!

File: sockjsclient/xhr.go (hunk 6 above)
Comment (fatih) [inline sockjsclient/xhr.go]: Done

**Pull request merged**